File: leetcode/predict-the-winner.py
from functools import lru_cache
from typing import List


class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        @lru_cache(maxsize=1000)
        def canBeAWinner(l: int, r: int, player1s_turn: bool) -> int:
            if l == r:
                return 0

            # take first or last
            score = max(
                nums[l] - canBeAWinner(l + 1, r, not player1s_turn),
                nums[r - 1] - canBeAWinner(l, r - 1, not player1s_turn),
            )

            return score

        return canBeAWinner(0, len(nums), True) >= 0

File: leetcode/predict-the-winner.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    bool PredictTheWinner(vector<int>& nums)
    {
        auto size = nums.size();

        vector<vector<int>> memo(size, vector<int>(size));

        function<int(int, int)> canWin = [&](int l, int r) {
            if (l == r) {
                return nums[l];
            }

            if (memo[l][r] == 0) {
                memo[l][r] = max(nums[l] - canWin(l + 1, r), nums[r] - canWin(l, r - 1));
            }

            return memo[l][r];
        };

        return canWin(0, size - 1) >= 0;
    }
};
